subsys/net/lib/nrf_cloud/coap/src/nrfc_dtls.c

```diff
@@ -0,0 +1,267 @@
+
+#include "nrfc_dtls.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(dtls, CONFIG_NRF_CLOUD_COAP_LOG_LEVEL);
+
+static int sectag = CONFIG_NRF_CLOUD_COAP_SEC_TAG;
+static bool dtls_cid_active;
+static bool cid_supported = true;
+
+#if defined(CONFIG_MODEM_INFO)
+static struct modem_param_info mdm_param;
+
+static int get_modem_info(void)
+{
+	int err;
+
+	err = modem_info_string_get(MODEM_INFO_IMEI,
+				    mdm_param.device.imei.value_string,
+				    MODEM_INFO_MAX_RESPONSE_SIZE);
+	if (err <= 0) {
+		LOG_ERR("Could not get IMEI: %d", err);
+		return err;
+	}
+
+	err = modem_info_string_get(MODEM_INFO_FW_VERSION,
+				    mdm_param.device.modem_fw.value_string,
+				    MODEM_INFO_MAX_RESPONSE_SIZE);
+	if (err <= 0) {
+		LOG_ERR("Could not get mfw ver: %d", err);
+		return err;
+	}
+
+	LOG_INF("IMEI:                    %s", mdm_param.device.imei.value_string);
+	LOG_INF("Modem FW version:        %s", mdm_param.device.modem_fw.value_string);
+
+	return 0;
+}
+
+#endif /* CONFIG_MODEM_INFO */
+
+static int get_device_ip_address(uint8_t *d4_addr)
+{
+#if defined(CONFIG_MODEM_INFO)
+	int err;
+
+	err = modem_info_string_get(MODEM_INFO_IP_ADDRESS,
+				    mdm_param.network.ip_address.value_string,
+				    MODEM_INFO_MAX_RESPONSE_SIZE);
+	if (err <= 0) {
+		LOG_ERR("Could not get IP addr: %d", err);
+		return err;
+	}
+	err = inet_pton(AF_INET, mdm_param.network.ip_address.value_string, d4_addr);
+	if (err == 1) {
+		return 0;
+	}
+	return -errno;
+#else
+	d4_addr[0] = 0;
+	d4_addr[1] = 0;
+	d4_addr[2] = 0;
+	d4_addr[3] = 0;
+	return 0;
+#endif
+
+}
+
+int nrfc_dtls_setup(int sock)
+{
+	int err;
+	uint8_t d4_addr[4];
+
+	/* once connected, cache the connection info */
+	dtls_cid_active = false;
+
+#if defined(CONFIG_MODEM_INFO)
+	err = get_modem_info();
+	if (err) {
+		LOG_INF("Modem firmware version not known");
+	}
+#endif
+
+	err = get_device_ip_address(d4_addr);
+	if (!err) {
+		LOG_DBG("Client IP address: %u.%u.%u.%u",
+			d4_addr[0], d4_addr[1], d4_addr[2], d4_addr[3]);
+	}
+
+	LOG_DBG("Setting socket options:");
+
+	LOG_DBG("  hostname: %s", CONFIG_NRF_CLOUD_COAP_SERVER_HOSTNAME);
+	err = setsockopt(sock, SOL_TLS, TLS_HOSTNAME, CONFIG_NRF_CLOUD_COAP_SERVER_HOSTNAME,
+			 sizeof(CONFIG_NRF_CLOUD_COAP_SERVER_HOSTNAME));
+	if (err) {
+		LOG_ERR("Error setting hostname: %d", -errno);
+		return -errno;
+	}
+
+	LOG_DBG("  sectag: %d", sectag);
+	err = setsockopt(sock, SOL_TLS, TLS_SEC_TAG_LIST, &sectag, sizeof(sectag));
+	if (err) {
+		LOG_ERR("Error setting sectag list: %d", -errno);
+		return -errno;
+	}
+
+	int cid_option = TLS_DTLS_CID_SUPPORTED;
+
+	cid_supported = true;
+	LOG_DBG("  Enable connection id");
+	err = setsockopt(sock, SOL_TLS, TLS_DTLS_CID, &cid_option, sizeof(cid_option));
+	if (!err) {
+	} else if ((err != EOPNOTSUPP) && (err != EINVAL)) {
+		LOG_ERR("Error enabling connection ID: %d", -errno);
+		cid_supported = false;
+	} else {
+		LOG_INF("Connection ID not supported by the installed modem firmware");
```

Comment:
nit: `Connection ID not supported by the provided socket` since this implementation really doesn't seem to care about whether there's a modem attached